File: Editorial/119_Pascal_s_Triangle_II/CSharp/2_Dynamic_Programming.cs
public class Solution {
    private Dictionary<(int, int), int> cache =
        new Dictionary<(int, int), int>();

    private int GetNum(int row, int col) {
        if (cache.ContainsKey((row, col))) {
            return cache[(row, col)];
        }

        int computedVal = (row == 0 || col == 0 || row == col)
                              ? 1
                              : GetNum(row - 1, col - 1) + GetNum(row - 1, col);
        cache[(row, col)] = computedVal;
        return computedVal;
    }

    public IList<int> GetRow(int rowIndex) {
        List<int> ans = new List<int>();
        for (int i = 0; i <= rowIndex; i++) {
            ans.Add(GetNum(rowIndex, i));
        }

        return ans;
    }
}

File: Editorial/0101-0200/0119_Pascal_s_Triangle_II/C/2_Dynamic_Programming.c
struct Cache {
    int key;
    int val;
    UT_hash_handle hh;
};
struct Cache *cache, *tmp;
void add_cache(int key, int val) {
    tmp = malloc(sizeof(struct Cache));
    tmp->key = key;
    tmp->val = val;
    HASH_ADD_INT(cache, key, tmp);
}
struct Cache* find_cache(int key) {
    HASH_FIND_INT(cache, &key, tmp);
    return tmp;
}
int getNum(int row, int col) {
    int key = ((row + 1) << 16) | (col + 1);
    tmp = find_cache(key);
    if (tmp) return tmp->val;
    if (row == 0 || col == 0 || row == col) {
        add_cache(key, 1);
        return 1;
    }
    int val = getNum(row - 1, col - 1) + getNum(row - 1, col);
    add_cache(key, val);
    return val;
}
int* getRow(int rowIndex, int* returnSize) {
    cache = NULL;
    *returnSize = rowIndex + 1;
    int* ans = malloc((*returnSize) * sizeof(int));
    for (int i = 0; i < *returnSize; i++) {
        ans[i] = getNum(rowIndex, i);
    }
    HASH_CLEAR(hh, cache);
    return ans;
}
